torchdata/datapipes/iter/transform/callable.py

```diff
@@ -168,6 +171,140 @@ def __len__(self) -> int:
+    and yields one item from this iterable (functional name: ``shuffled_flatmap``).
+
+    Note:
+        The output from ``fn`` must be an Iterable. Otherwise, an error will be raised.
+        If ``fn`` is ``None``, source DataPipe will be just flattened vertically, provided that items can be unpacked.
+
+    Args:
+        datapipe: Source IterDataPipe
+        fn: the function to be applied to each element in the DataPipe, the output must be a Sequence
+        input_col: Index or indices of data which ``fn`` is applied, such as:
+
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is/are used for list/tuple.
+            - Key(s) is/are used for dict.
+        buffer_size: The buffer size for sampling the iterables (default to ``10000``)
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> source_dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
+        >>> shuffled_flatmapped_dp = source_dp.shuffled_flatmap(buffer_size=2)
+        >>> list(shuffled_flatmapped_dp)
+        ['a', 'b', 'c', 1, 'd', 'A', 'B', 'C', 2, 'D', 3, 4]
+    """
+    datapipe: IterDataPipe[T_co]
+    fn: Optional[Callable]
+    buffer_size: int
+    _enabled: bool
+    _seed: Optional[int]
+    _rng: random.Random
+
+    def __init__(
+        self, datapipe: IterDataPipe, fn: Optional[Callable] = None, input_col=None, buffer_size: int = 10000
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+
+        if fn is None:
+            fn = _no_op_fn
+        _check_unpickable_fn(fn)
+        self.fn = fn  # type: ignore[assignment]
+        self.input_col = input_col
+        validate_input_col(fn, input_col)
+
+        assert buffer_size > 0, "buffer_size should be larger than 0"
+        self.buffer_size = buffer_size
+        self._enabled = True
+        self._seed = None
+        self._rng = random.Random()
+
+    def set_shuffle(self, shuffle=True):
+        self._enabled = shuffle
+        return self
+
+    def set_seed(self, seed: int):
+        self._seed = seed
+        return self
+
+    def reset(self) -> None:
+        if self._enabled:
+            if self._seed is None:
+                self._seed = int(torch_empty((), dtype=torch_int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+
+    def _apply_fn(self, data):
+        if self.input_col is None:
+            return self.fn(data)  # type: ignore[misc]
+        elif isinstance(self.input_col, (list, tuple)):
+            args = tuple(data[col] for col in self.input_col)
+            return self.fn(*args)  # type: ignore[misc]
+        else:
+            return self.fn(data[self.input_col])  # type: ignore[misc]
+
+    def _yield_one(self, buffer: List[Iterator[T_co]]) -> Iterator[T_co]:
+        idx = self._rng.randint(0, len(buffer) - 1) if self._enabled else 0
+        try:
+            yield next(buffer[idx])
```

Comment:
Don't you think that replacing the `next` by a random choice would overcomplicate the class?
To get a full shuffle, I think I would simply do:
```python
dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
dp = dp.map(IterableWrapper)
dp = dp.map(Shuffler)
dp = dp.shuffled_flatmap()
list(dp)
# ['b', 3, 'c', 'd', 'C', 'A', 'a', 2, 'B', 'D', 4, 1]
```

Reply:
I used a `yield` to avoide a case distinction for when an iterator is empty or not in the `__iter__` function bellow and reduce the amount of code. But we can use `return` indeed

Reply:
> ```python
> dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
> dp = dp.map(IterableWrapper)
> dp = dp.map(Shuffler)
> dp = dp.shuffled_flatmap()
> ```

Emmm. This is not correct. This `shuffled_flatmap` is helped to shuffle across input object. We still need to shuffle the inner object for each input. This might work though
```py
dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
dp = dp.in_batch_shuffle()
dp = dp.shuffled_flatmap()
```

Reply:
I added an example of how to completely shuffle the input in the doctoring: https://github.com/PierreGtch/TorchData/commit/cf887655a150235428b5af7aa4b7a4469bbd1573 

Is that enough?